packages/mwp-router/src/util/resolve.js

```diff
@@ -1,119 +1,54 @@
 	}
-	return _resolveNestedRoutes(matchedRoute).then(m => m.route.routes || []);
+	return route.getComponent();
 };
 
-const _resolveNestedRoutes: Resolver<MatchedRoute> = matchedRoute =>
-	matchedRoute.route.getNestedRoutes
-		? matchedRoute.route
-				.getNestedRoutes()
-				.then(routes => (matchedRoute.route.routes = routes))
-				.then(() => matchedRoute)
-		: Promise.resolve(matchedRoute);
-
-const _resolveIndexRoute: Resolver<MatchedRoute> = matchedRoute =>
-	matchedRoute.route.getIndexRoute
-		? matchedRoute.route
-				.getIndexRoute()
-				.then(indexRoute => (matchedRoute.route.indexRoute = indexRoute))
-				.then(() => matchedRoute)
-		: Promise.resolve(matchedRoute);
-
 /*
- * Find all routes in the routes array that match the provided URL, including
- * nested routes that may be async
+ * Resolve the current route and all children
+ * 
+ * *Note* DO NOT USE THIS IN THE BROWSER - it will eagerly resolve all async
+ * components
  */
-const _resolveRouteMatches = (
-	routes: Array<PlatformRoute> = [],
-	path: string = '',
-	matchedRoutes: Array<MatchedRoute> = [],
-	matchedPath: string = ''
-): Promise<Array<MatchedRoute>> => {
-	const route = routes.find(r =>
-		matchPath(path, _routeMatchOptions(r, matchedPath))
-	); // take the first match
-	if (!route) {
-		return Promise.resolve(matchedRoutes);
-	}
-
-	// add the route and its `match` object to the array of matched routes
-	const currentMatchOptions = _routeMatchOptions(route, matchedPath);
-	const match = matchPath(path, currentMatchOptions);
-	if (!match) {
-		// we know that this won't ever run because we've established the match in
-		// `.find`, but this check is for type safety
-		return Promise.resolve(matchedRoutes);
-	}
-	const matchedRoute = { route, match };
-	const currentMatchedRoutes = [...matchedRoutes, matchedRoute];
-
-	// add any nested route matches
-	return resolveChildRoutes(matchedRoute).then(
-		childRoutes =>
-			childRoutes.length
-				? _resolveRouteMatches(
-						childRoutes,
-						path,
-						currentMatchedRoutes,
-						currentMatchOptions.path
-				  )
-				: currentMatchedRoutes
+export const resolveRoute = (
+	route: PlatformRoute
+): Promise<StaticPlatformRoute> =>
+	Promise.all([
+		resolveComponent(route),
+		// $FlowFixMe - Flow doesn't realize the returned promise will be unwrapped
+		resolveAllRoutes(route.routes || []),
+	]).then(
+		(
+			[component: React$ComponentType<*>, routes: Array<StaticPlatformRoute>]
+		): StaticPlatformRoute =>
+			Object.freeze({
+				...addComponentToRoute(route)(component),
+				routes,
+			})
 	);
-};
 
-/*
- * An curried interface into `_resolveRouteMatches`, using `basename`
- * + `location` instead of `path`
- */
-export const getRouteResolver = (routes: Array<PlatformRoute>, basename: string) => (
-	location: URL
-): Promise<Array<MatchedRoute>> => {
-	const path = location.pathname.replace(basename, '');
-	return _resolveRouteMatches(routes, path);
-};
+export const resolveAllRoutes = (
```

Comment:
not sure if it's really worth implementing this as a standalone function since the implementation is so simple - parent could call `Promise.all(routes.map(resolveRoute));` itself pretty easily...